hazelcast-client/src/test/java/com/hazelcast/client/cache/nearcache/ClientNearCacheInvalidationTest.java

```diff
@@ -0,0 +1,355 @@
+            @Override
+            public void run()
+                    throws Exception {
+                assertEquals(1, counter.get());
+            }
+        }, TIMEOUT);
+    }
+
+    @Test
+    public void when_cacheCleared_invalidationEventIsReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        clear();
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run()
+                    throws Exception {
+                assertEquals(1, counter.get());
+            }
+        }, TIMEOUT);
+    }
+
+    @Test
+    public void when_cacheClosed_invalidationEventIsNotReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        close();
+        assertTrueAllTheTime(new AssertTask() {
+            @Override
+            public void run()
+                    throws Exception {
+                assertEquals(0, counter.get());
+            }
+        }, TIMEOUT);
+    }
+
+    // When CacheManager.destroyCache is invoked from client-side CacheManager, an invalidation event is received
+    // When invoked from a member-side CacheManager, invalidation event is not received.
+    @Test
+    public void when_cacheManagerDestroyCacheInvoked_invalidationEventMayBeReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        destroyCacheFromCacheManager();
+        if (invokeCacheOperationsFromMember) {
+            // when from a member-side cache manager, invalidation event cannot be received, as the invalidation listener
+            // registration is removed *before* the invocation to destroy the cache is sent to the member
+            assertTrueAllTheTime(new AssertTask() {
+                @Override
+                public void run()
+                        throws Exception {
+                    assertEquals(0, counter.get());
+                }
+            }, TIMEOUT);
+        } else {
+            // when from a client-side cache manager, an invalidation event is received
+            assertTrueEventually(new AssertTask() {
+                @Override
+                public void run()
+                        throws Exception {
+                    assertEquals(1, counter.get());
+                }
+            }, TIMEOUT);
+        }
+    }
+
+    @Test
+    public void when_shuttingDown_invalidationEventIsNotReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        shutdown();
+        assertTrueAllTheTime(new AssertTask() {
+            @Override
+            public void run()
+                    throws Exception {
+                assertEquals(0, counter.get());
+            }
+        }, TIMEOUT);
+    }
+
+    private void registerInvalidationListener(AtomicInteger counter) {
```

Comment:
We only care about the client side in this test, there is no member-side near cache for JCache in 3.7.